File: Source/VRTAsPrimary/VRTInterstitialCustomEventIronSource.m
//
//  VRTInterstitialCustomEventIronSource.m
//
//  Created by Scott McCoy on 5/9/19.
//  Copyright © 2019 VRTCAL. All rights reserved.
//

//Header
#import "VRTInterstitialCustomEventIronSource.h"

//Dependencies
#import "VRTIronSourceManager.h"
#import <IronSource/IronSource.h>

//IronSource Interstitial Adapter, Vrtcal as Primary
@interface VRTInterstitialCustomEventIronSource() <ISInterstitialAdDelegate>
@property NSString *instanceId;
@end


@implementation VRTInterstitialCustomEventIronSource

- (void) loadInterstitialAd {
    
    @try {

        //Get and validate app ID
        NSString *appKey = [self.customEventConfig.thirdPartyCustomEventData objectForKey:@"appId"];
        if (![self stringIsGood:appKey]) {
            VRTError *vrtError = [VRTError errorWithCode:VRTErrorCodeInvalidParam format:@"IronSource appId of %@ unusable", appKey];
            [self.customEventLoadDelegate customEventFailedToLoadWithError:vrtError];
            return;
        }
        
        //Init IronSource if it hasn't been already
        [[VRTIronSourceManager singleton] initIronSourceSDKWithAppKey:appKey forAdUnits:[NSSet setWithObject:IS_INTERSTITIAL]];
        
        //Check if an interstitial is ready
        if ([IronSource hasInterstitial]) {
            [self.customEventLoadDelegate customEventLoaded];
            return;
        }
        
        //Request an interstitial
        [IronSource setInterstitialDelegate:self];
        [IronSource loadInterstitial];
    } @catch (NSException *exception) {
        
        NSString *description = [exception description];
        VRTError *vrtError = [VRTError errorWithCode:VRTErrorCodeUnknown format:@"Exception in loadInterstitialAd: %@", description];
        [self.customEventLoadDelegate customEventFailedToLoadWithError:vrtError];
    }

}

- (void) showInterstitialAd {
    UIViewController *vc = [self.viewControllerDelegate vrtViewControllerForModalPresentation];
    [IronSource showInterstitialWithViewController:vc];
}



#pragma mark - ISInterstitialDelegate
- (void)didClickInterstitial {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventClicked];
}

- (void)interstitialDidClose {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventDidDismissModal:VRTModalTypeInterstitial];
}

- (void)interstitialDidFailToLoadWithError:(NSError *)error {
    VRTLogWhereAmI();
    [self.customEventLoadDelegate customEventFailedToLoadWithError:error];
}

- (void)interstitialDidFailToShowWithError:(NSError *)error {
    //No VRT analog for this
    VRTLogWhereAmI();
}

- (void)interstitialDidLoad {
    VRTLogWhereAmI();
    [self.customEventLoadDelegate customEventLoaded];
}

- (void)interstitialDidOpen {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventWillPresentModal:VRTModalTypeInterstitial];
}

- (void)interstitialDidShow {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventDidPresentModal:VRTModalTypeInterstitial];
}


#pragma mark - Utility
- (BOOL) stringIsGood:(NSString*)str {
    return str != nil && [str isKindOfClass:[NSString class]] && [str length] > 0;
}

@end

File: Source/VRTAsPrimary/VRTInterstitialCustomEventIronSource.swift

//
//  VRTInterstitialCustomEventIronSource.swift
//
//  Created by Scott McCoy on 5/9/19.
//  Copyright © 2019 VRTCAL. All rights reserved.
//

import VrtcalSDK
import IronSource

//IronSource Interstitial Adapter, Vrtcal as Primary

class VRTInterstitialCustomEventIronSource: VRTAbstractInterstitialCustomEvent, ISInitializationDelegate {
    static var ironSourceInitialized = false
    var levelPlayInterstitialDelegatePassthrough = LevelPlayInterstitialDelegatePassthrough()
    
    override func loadInterstitialAd() {
        VRTLogInfo()
        
        // Bail if IS is already initialized
        guard !Self.ironSourceInitialized else {
            finishLoadingInterstitial()
            return
        }
        
        // Get and validate app ID
        guard let appKey = customEventConfig.thirdPartyCustomEventDataValueOrFailToLoad(
            thirdPartyCustomEventKey: ThirdPartyCustomEventKey.appId,
            customEventLoadDelegate: customEventLoadDelegate
        ) else {
            return
        }
        
        // Set the mediation type
        let mediationName = "vrtcal"
        let isMediationVersion = "500"
        let vrtcalSdkVersion = VrtcalSDK.sdkVersion()
        let mediationType = "\(mediationName)\(isMediationVersion)SDK\(vrtcalSdkVersion)"
        IronSource.setMediationType(mediationType)

        IronSource.initWithAppKey(
            appKey,
            adUnits: [IS_INTERSTITIAL],
            delegate: self
        )
        
        Self.ironSourceInitialized = true
    }

    override func showInterstitialAd() {
        VRTLogInfo()
        
        levelPlayInterstitialDelegatePassthrough.customEventShowDelegate = customEventShowDelegate
        
        guard let vc = viewControllerDelegate?.vrtViewControllerForModalPresentation() else {
            customEventShowDelegate?.customEventFailedToShow(vrtError: .customEventViewControllerNil)
            return
        }
        IronSource.showInterstitial(with: vc)
    }
    
    func finishLoadingInterstitial() {
        VRTLogInfo()

        // Check if an interstitial is ready
        if IronSource.hasInterstitial() {
            VRTLogInfo("IronSource has interstitial, bailing")
            customEventLoadDelegate?.customEventLoaded()
            return
        }

        // Request an interstitial
        levelPlayInterstitialDelegatePassthrough.customEventLoadDelegate = customEventLoadDelegate
        levelPlayInterstitialDelegatePassthrough.customEventShowDelegate = customEventShowDelegate
        
        IronSource.setLevelPlayInterstitialDelegate(levelPlayInterstitialDelegatePassthrough)
        IronSource.loadInterstitial()
    }

    func initializationDidComplete() {
        VRTLogInfo()
        finishLoadingInterstitial()
    }
}


